Introduce support for automated log rotation/backup in pkg/log.

geeknoid opened the pull request:
This adds a few cmd-line options to support this feature. You can
specify max log size before rotation, max log age to keep, and max #
of logs to keep.

This package has 100% test code coverage.

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 7080dc308cd167d4c07688a74ba52b66cd121336 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2313/istio-presubmit/1698/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | fc3f381ecec74bc7b77a9fa0f19ed96ca7ef8090 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2313/istio-presubmit/1699/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

geeknoid commented:
/retest all

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | fc3f381ecec74bc7b77a9fa0f19ed96ca7ef8090 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2313/istio-presubmit/1704/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

nmittler commented on pkg/log/options.go:
```diff
@@ -24,15 +24,50 @@ import (
 	// OutputPaths is a list of file system paths to write the log data to.
 	// The special values stdout and stderr can be used to output to the
-	// standard I/O streams.
+	// standard I/O streams. THis defaults to stdout.
```
nit: `This`...

geeknoid replied:
Fkxed

nmittler commented on pkg/log/options.go:
```diff
@@ -24,15 +24,50 @@ import (
 
+	// ErrorOutputPaths is a list of file system paths to write logger errors to.
+	// The special values stdout and stderr can be used to output to the
+	// standard I/O streams. THis defaults to stderr.
```
`This` again

geeknoid replied:
Fixed, along with a few other comment errors I noticed.

nmittler commented on pkg/log/options.go:
```diff
@@ -24,15 +24,50 @@ import (
 	// OutputPaths is a list of file system paths to write the log data to.
 	// The special values stdout and stderr can be used to output to the
-	// standard I/O streams.
+	// standard I/O streams. THis defaults to stdout.
```
Does it make sense to do rotation if you're using stdout/stderr? I would think that the list should either have a single element `stdout` or have a list of files, but not both. Would it make more sense to eliminate the use of `stdout` altogether and just assume `stdout` as the default if the list is empty?   The same would be true for `ErrorOutputPaths` with `stderr`.

geeknoid replied:
The default is output to stdout with no log rotation.

If you use log rotation, you must supply explicit output paths. The first such output path cannot be stdout and will be the file that gets rotated. Other output paths will be written to without rotation and can include stdout.

I could simplify things if we want. The multiple output paths thing may be overkill (zap supports it inherently, that's why I exposed it). We could go instead with:

- A boolean to control output to stdout
- A single rotation output path.

Or perhaps:

- A plain single output path. This could be a filename, stdout, or stderr
- A rotation output path. This would be a filename.

WDYT?

nmittler replied:
Oh, I must have read through the code too quickly (apologies) ... I didn't know that log rotation only operated on a single file.  I assumed that it would round-robin across the files in the list, overwriting as necessary.

Let me make sure I understand what we want to do:
- Allow user to specify a log file to be rotated
  - When it's full, future writes will result in losing earliest data.
  - stdout/stderr should never be used
- Allow user to specify at least one non-rotating log file (e.g. stdout)
  - default is stdout/stderr

I would think that maybe the special bit here is the rotating support.  I think what you have for `OutputPaths` is fine, but maybe they shouldn't be involved in rotation at all.  Perhaps we should allow the user to specify a `RotatingOutputPath` (or `RotatingLogFile` since it's specifically a file and not a directory) to enable rotation?

WDYT?

geeknoid replied:
First, a correction. Rotation doesn't generally destroy the log file. It renames the current log file by inserting a timestamp into the name and then starts writing to the original name again. You can configure how many of these backups to keep and/or how long to keep old ones.

Second, I think what you're suggesting was my second option from above. A specific path for output (a file, stdout, or stderr) and a specific distinct path for rotation (always a file). Is that right?

nmittler replied:
> First, a correction. Rotation doesn't generally destroy the log file. It renames the current log file by inserting a timestamp into the name and then starts writing to the original name again. You can configure how many of these backups to keep and/or how long to keep old ones.

Ok that makes sense.  So for rotation we'll specify a directory (and maybe a log file prefix), correct? (since we'll be creating files as we go)

> Second, I think what you're suggesting was my second option from above. A specific path for output (a file, stdout, or stderr) and a specific distinct path for rotation (always a file). Is that right?

Yes, basically.  The only difference is that I think the non-rotated paths should allow multiple .. not just one.  That way we can log to multiple sources at once (stdout and a file).

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | de1185467f0c2019209a69039d0da144d1773951 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2313/istio-presubmit/1718/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

geeknoid commented:
/retest all

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | de1185467f0c2019209a69039d0da144d1773951 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2313/istio-presubmit/1719/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

geeknoid commented:
/retest all

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | de1185467f0c2019209a69039d0da144d1773951 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2313/istio-presubmit/1720/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 24e9242f4c650e7eec719035b14866fd41f152eb | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2313/istio-presubmit/1721/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 5c49ff3561b17d04c52a8eed11d85c769a69a672 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2313/istio-presubmit/1722/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

geeknoid commented:
OK, I'll adjust the options accordingly.

Note that for the rotation file, you don't give it a directory. You give it
a base filename, and it generates filenames in the same directory by
tacking on timestamps. That's how the lumberjack lib I'm using does it.

Anyway, I'll update the options.

On Fri, Dec 22, 2017 at 9:45 AM, Nathan Mittler <notifications@github.com>
wrote:

> *@nmittler* commented on this pull request.
> ------------------------------
>
> In pkg/log/options.go
> <https://github.com/istio/istio/pull/2313#discussion_r158531381>:
>
> >  )
>
>  // Options defines the set of options supported by Istio's component logging package.
>  type Options struct {
>  	// OutputPaths is a list of file system paths to write the log data to.
>  	// The special values stdout and stderr can be used to output to the
> -	// standard I/O streams.
> +	// standard I/O streams. THis defaults to stdout.
>
> First, a correction. Rotation doesn't generally destroy the log file. It
> renames the current log file by inserting a timestamp into the name and
> then starts writing to the original name again. You can configure how many
> of these backups to keep and/or how long to keep old ones.
>
> Ok that makes sense. So for rotation we'll specify a directory (and maybe
> a log file prefix), correct? (since we'll be creating files as we go)
>
> Second, I think what you're suggesting was my second option from above. A
> specific path for output (a file, stdout, or stderr) and a specific
> distinct path for rotation (always a file). Is that right?
>
> Yes, basically. The only difference is that I think the non-rotated paths
> should allow multiple .. not just one. That way we can log to multiple
> sources at once (stdout and a file).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/2313#discussion_r158531381>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AVucHdcP3Ogeu0_HQnrdkBSFcs8QAun7ks5tC-rNgaJpZM4RKVUy>
> .
>

nmittler commented:
> Note that for the rotation file, you don't give it a directory. You give it
a base filename, and it generates filenames in the same directory by
tacking on timestamps. That's how the lumberjack lib I'm using does it.

Ok that works too ... maybe make that clear with the property name.

Are you happy with the suggestion?  Don't just implement to make me happy :)

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 2fe93b1de0d62b45a30f681bb9d329cd51bdf8cb | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2313/istio-presubmit/1748/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

geeknoid commented:
PTAL.

Now treats rotated file output as an orthogonal notion to the main output file paths.
I like this much better than what I had.

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | e4e3ddda8c682d727dbe266c0bb5a2febd84cefb | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2313/istio-presubmit/1750/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

nmittler commented on pkg/log/options.go:
```diff
@@ -24,15 +24,48 @@ import (
+	// RotateOutputPath is the path to a rotating log file. This file should
+	// be automatically rotated over time, based on the rotation parameters such
+	// as RotationMaxSize and RotationMaxAge. The default is to not rotate.
+	RotateOutputPath string
```
I like this much better! :)

The only thing that's still a little unclear to me is how log file naming should work, based on these options (and the supporting comments). The fact that this is an `OutputPath` and not an `OutputDir` leads me to initially think that there will only be a single rotating log file generated (which I name explicitly here). The other parameters (max size, age, etc.), however, suggest that multiple files will be generated by the rotation process.  Perhaps it would make more sense to make this a directory rather than a file? Or was that your intent?

geeknoid commented:
Not quite right. The path is where logs are output. When we hit a max age
or max size, then the file is renamed by tacking on a timestamp to it, and
then output resumes to the original file path. I've updated the comment in
the code to read:

// RotateOutputPath is the path to a rotating log file. This file should
// be automatically rotated over time, based on the rotation parameters such
// as RotationMaxSize and RotationMaxAge. The default is to not rotate.
//
// This path is used as a foundational path. This is where log output is
normally
// saved. When a rotation needs to take place because the file got too big
or too
// old, then the file is renamed by appending a timestamp to the name. Such
renamed
// files are called backups. Once a backup has been created,
// output resumes to this path.


On Wed, Dec 27, 2017 at 8:03 AM, Nathan Mittler <notifications@github.com>
wrote:

> *@nmittler* commented on this pull request.
> ------------------------------
>
> In pkg/log/options.go
> <https://github.com/istio/istio/pull/2313#discussion_r158833611>:
>
> >  	OutputPaths []string
>
> +	// ErrorOutputPaths is a list of file system paths to write logger errors to.
> +	// The special values stdout and stderr can be used to output to the
> +	// standard I/O streams. This defaults to stderr.
> +	ErrorOutputPaths []string
> +
> +	// RotateOutputPath is the path to a rotating log file. This file should
> +	// be automatically rotated over time, based on the rotation parameters such
> +	// as RotationMaxSize and RotationMaxAge. The default is to not rotate.
> +	RotateOutputPath string
>
> I like this much better! :)
>
> The only thing that's still a little unclear to me is how log file naming
> should work, based on these options (and the supporting comments). The fact
> that this is an OutputPath and not an OutputDir leads me to initially
> think that there will only be a single rotating log file generated (which I
> name explicitly here). The other parameters (max size, age, etc.), however,
> suggest that multiple files will be generated by the rotation process.
> Perhaps it would make more sense to make this a directory rather than a
> file? Or was that your intent?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/2313#pullrequestreview-85695196>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AVucHVpRma2QPdp26rQ2JVFkQFeJaY7aks5tEmpbgaJpZM4RKVUy>
> .
>
